File: 47 Permutations II/Python/Solution.py
from typing import List, Self


class Solution:
    def backtrack(self: Self, nums: List[int], permutations: List[List[int]], permutation: List[int], consumed: List[int]) -> None:
        l: int = len(nums)
        if len(permutation) == l:
            if permutation not in permutations:
                permutations.append(permutation.copy())
            return
        
        for i in range(l):
            if i not in consumed:
                permutation.append(nums[i])
                consumed.append(i)
                self.backtrack(nums, permutations, permutation, consumed)
                consumed.pop()
                permutation.pop()
        
    def permuteUnique(self: Self, nums: List[int]) -> List[List[int]]:
        permutations: List[List[int]] = []
        self.backtrack(nums, permutations, [], [])
        return permutations

File: 47 Permutations II/Java/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    private static void backtrack(int[] nums, List<List<Integer>> permutations, List<Integer> permutation, List<Integer> consumed) {
        if (permutation.size() == nums.length) {
            if (!permutations.contains(permutation)) {
                permutations.add(new ArrayList<>(permutation));
            }
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (!consumed.contains(i)) {
                permutation.add(0, nums[i]);
                consumed.add(0, i);
                backtrack(nums, permutations, permutation, consumed);
                permutation.remove(0);
                consumed.remove(0);
            }
        }
    }

    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> permutations = new ArrayList<>();
        backtrack(nums, permutations, new ArrayList<>(), new ArrayList<>());
        return permutations;
    }
}
